### PR TITLE
python3Packages.slack-bolt: init at 1.16.1

### PR DESCRIPTION
**Pull request by jpetrucciani**
###### Description of changes

I would like to include [`slack-bolt`](https://github.com/slackapi/bolt-python/) in nixpkgs to be able to create [Slack](https://slack.com/) bots and other applications using Slack's [Bolt](https://api.slack.com/tools/bolt) framework! This PR also includes adding [`boddle`](https://github.com/keredson/boddle) as a `checkInput`.

Previous PR for `boddle` here: #188832

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/boddle/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+, buildPythonPackage
+, fetchFromGitHub
+, pythonOlder
+, bottle
+}:
+
+buildPythonPackage rec {
+  pname = "boddle";
+  version = "0.2.9";
+  disabled = pythonOlder "3.7";
```
Comment (fabaff) [inline pkgs/development/python-modules/boddle/default.nix:11]: Not required at the moment but perhaps in the future.

```suggestion
  version = "0.2.9";
  format = "setuptools";

  disabled = pythonOlder "3.7";
```

File: pkgs/development/python-modules/boddle/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  src = fetchFromGitHub {
+    owner = "keredson";
+    repo = pname;
+
+    # upstream doesn't tag releases
```
Comment (fabaff) [inline pkgs/development/python-modules/boddle/default.nix:17]: Perhaps we get [tags](https://github.com/keredson/boddle/issues/14).

File: pkgs/development/python-modules/boddle/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  propagatedBuildInputs = [ bottle ];
+
+  pythonImportsCheck = [
+    "boddle"
+  ];
+
+  meta = with lib; {
+    description = "Unit testing tool for Python's bottle library";
+    homepage = "https://github.com/keredson/boddle";
+    license = licenses.lgpl21;
```
Comment (fabaff) [inline pkgs/development/python-modules/boddle/default.nix:31]: The `setup.py` says [LGPLv3](https://github.com/keredson/boddle/blob/master/setup.py#L27) and the LICENSE file [LGPLv2.1](https://github.com/keredson/boddle/blob/master/LICENSE).

```suggestion
    license = licenses.lgpl21Only;
```

https://github.com/keredson/boddle/issues/15

File: pkgs/development/python-modules/slack-bolt/default.nix [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  src = fetchFromGitHub {
+    owner = "slackapi";
+    repo = "bolt-python";
+    rev = "v${version}";
+    sha256 = "sha256-xzEsTx+44ae++WxArxO2rhCsot+ELuhc0aZ44MXY0c4=";
+  };
+
+  propagatedBuildInputs = [ slack-sdk ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/slack-bolt/default.nix:40]: Use `optional-dependencies` for extra requirements.

```suggestion

  passthru.optional-dependencies = {
    async = [
      aiohttp
      websockets
    ];
   adapter  = [
      boto3
      ...
    ];
  };

```

And then for the check inputs: 

```nix
  checkInputs = [
    pytestCheckHook
    ...
  ] ++ passthru.optional-dependencies.adapter;
```

File: pkgs/development/python-modules/slack-bolt/default.nix [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    "test_oauth"
+  ];
+
+  # patch out pytest-runner
+  preBuild = ''
+    sed -i '/pytest-runner/d' ./setup.py
+  '';
+
+  meta = with lib; {
+    description = "A framework to build Slack apps using Python ";
```
Comment (fabaff) [inline pkgs/development/python-modules/slack-bolt/default.nix:96]: More of an opinionated change, expect the whitespace which should be removed at the end.

```suggestion
    description = "A framework to build Slack apps";
```

**Comment by jpetrucciani** [comment]
@fabaff thanks for the thorough review! I've made all the requested changes, and added a comment noting your issue in the boddle github repo!

**Comment by jpetrucciani** [comment]
This has switched `checkInputs` to `nativeCheckInputs` to satisfy the recent tree-wide change!

@fabaff is the lack of `boddle` tags a blocker for this PR in general? or would this be mergeable without those?

**Comment by jpetrucciani** [comment]
closing since #243593 was merged.

**Pull request closed without merging**